### PR TITLE
Creating eeprom_parition

### PR DESCRIPTION
**Pull request by bharath-v-m**
## Changes

Included function for taking in array for structures for creating a partition

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please reach out to your Project Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)
#70

### CODE REVIEW

File: middleware/src/eepromdirectory.c [hunk 1]
```diff
@@ -1,6 +1,38 @@
...
+  1. id of each fault
+  2. size of the fault
+ */
+struct eeprrom_struct
+{
+   char id;
+   int size;
+}
+
+/* taking in the number of user */
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:15]: these objects dont need to be defined here - the user has to pass them in from their application

File: middleware/src/eepromdirectory.c [hunk 2]
```diff
@@ -1,6 +1,38 @@
 #include "eepromdirectory.h"
 #include "general/include/m24c32.h"
 
+
+/*   user defined data type for taking in
+  1. id of each fault
+  2. size of the fault
+ */
+struct eeprrom_struct
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:9]: This struct is already defined in the .h file - no need to define here

File: middleware/src/eepromdirectory.c [hunk 3]
```diff
@@ -1,6 +1,38 @@
...
+int size_of_array = <size of the array>;
+
+/*  creating an array of struct(eeprom_input) for taking in input */
+int *eeprom_input = new eeprom_struct(size_of_array);
+
+
+/* function definition for entering parameters of the fault coniditon
+passing the array of struct to the function */
+
+void eepromInit(int* eeprom_input, int& array_size)
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:25]: this should be a pointer to the struct, not an int. Also, the array size will jsut be an int, no need for int&

File: middleware/src/eepromdirectory.c [hunk 4]
```diff
@@ -1,6 +1,38 @@
...
+{
+    for (int i = 0; i < array_size; i++)
+    {
+    eeprom_input[i]->id = <id>;
+    eeprom_input[i]->size = <size>;
+
+    // calculating the address through the logic defined in shepherd
+    }
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:33]: looks like there is a comment here showing the address calc, but doesnt look like its implemented.

**Review by dyldonahue — COMMENTED** [review]
Good start! definitely have the right idea, just a few things to fix up

**Review by dyldonahue — CHANGES_REQUESTED** [review]
Let me know if any of my comments dont make any sense

File: middleware/src/eepromdirectory.c (hunk 1 above)
Comment (bharath-v-m) [reply middleware/src/eepromdirectory.c:15]: rectified.

File: middleware/src/eepromdirectory.c (hunk 2 above)
Comment (bharath-v-m) [reply middleware/src/eepromdirectory.c:9]: removed it.

File: middleware/src/eepromdirectory.c (hunk 3 above)
Comment (bharath-v-m) [reply middleware/src/eepromdirectory.c:25]: Understood. Changed it accordingly.

File: middleware/src/eepromdirectory.c [hunk 5]
```diff
@@ -6,14 +6,6 @@
...
-{
-   char id;
-   int size;
-}
-
-/* taking in the number of user */
-int size_of_array = <size of the array>;
 
 /*  creating an array of struct(eeprom_input) for taking in input */
 int *eeprom_input = new eeprom_struct(size_of_array);
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:11]: this will come from the application, so we dont need to create the eeprom struct here. This file is only interface that applications use, so the input data will come from the applications. So, this line can be removed. 

(also side note: this is c++ syntax, in c, we have to use malloc() to dynamically allocate memory

File: middleware/src/eepromdirectory.c [hunk 6]
```diff
@@ -22,14 +14,20 @@ int *eeprom_input = new eeprom_struct(size_of_array);
...
-void eepromInit(int* eeprom_input, int& array_size)
+void eepromInit(struct* eeprom_input, int array_size)
 {
     for (int i = 0; i < array_size; i++)
     {
     eeprom_input[i]->id = <id>;
     eeprom_input[i]->size = <size>;
 
     // calculating the address through the logic defined in shepherd
+    while (eeprom_data[i].id != NULL)
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:25]: This is my fault because i didn't realize this until seeing them side by side - but this while loop can actually be incorporated directly into the for loop. Unlike the way it was in shepherd, this does not need to be a second, separate loop. Try to add the functionality of this while loop into the for loop above it

File: middleware/src/eepromdirectory.c [hunk 7]
```diff
@@ -1,6 +1,36 @@
...
+ */
+
+/*  creating an array of struct(eeprom_input) for taking in input */
+int *eeprom_input = new eeprom_struct(size_of_array);
+
+
+/* function definition for entering parameters of the fault coniditon
+passing the array of struct to the function */
+
+void eepromInit(struct* eeprom_input, int array_size)
```
Comment (dyldonahue) [inline middleware/src/eepromdirectory.c:17]: the syntax for this parameter is not quite right. the struct is called eeprom_partition. see definition in the .h file. We want a pointer to this datatype passed in

**Review by dyldonahue — CHANGES_REQUESTED** [review]
getting closer - a few small changes and this should be good to go

File: middleware/src/eepromdirectory.c (hunk 6 above)
Comment (bharath-v-m) [reply middleware/src/eepromdirectory.c:25]: I might be wrong here again, but i guess in the version I've uploaded the while condition for the address is in the for loop

File: middleware/include/eepromdirectory.h [hunk 8]
```diff
@@ -41,9 +42,12 @@ char *eeprom_get_key(int index);
  * @param key
  * @param data
  */
+
+/* declaration of function taking in array of structures for filling in the parameters of the fault conditions*/
+void eepromInit(int* eeprom_input, int& array_size)
```
Comment (nwdepatie) [inline middleware/include/eepromdirectory.h:47]: nit: incorrect naming convention, check other functions

Comment (nwdepatie) [inline middleware/include/eepromdirectory.h:47]: Will not build, & is not valid in C parameters iirc

File: middleware/include/eepromdirectory.h [hunk 9]
```diff
@@ -41,9 +42,12 @@ char *eeprom_get_key(int index);
...
 void eeprom_read_data_key(char *key, void *data);
 
-void eeprom_read_data_index(uint8_t index, void *data);
+void eeprom_read_data_index(uint8_t index, void *data);
```
Comment (nwdepatie) [inline middleware/include/eepromdirectory.h:50]: nit: added whitespace

File: middleware/src/eepromdirectory.c [hunk 10]
```diff
@@ -1,6 +1,33 @@
...
+ */
+
+
+/* function definition for entering parameters of the fault coniditon
+passing the array of struct to the function */
+
+void eepromInit(struct* eeprom_partition, int array_size)
+{
+    for (int i = 0; i < array_size; i++)
+    {
```
Comment (nwdepatie) [inline middleware/src/eepromdirectory.c:17]: nit: weird formatting, check other functions for curly brackets

File: middleware/src/eepromdirectory.c [hunk 11]
```diff
@@ -1,6 +1,33 @@
...
+    eeprom_input[i]->id = <id>;
```
Comment (nwdepatie) [inline middleware/src/eepromdirectory.c:18]: Will not build, not sure on why <> was used

File: middleware/src/eepromdirectory.c [hunk 12]
```diff
@@ -1,6 +1,33 @@
...
+    eeprom_input[i]->size = <size>;
```
Comment (nwdepatie) [inline middleware/src/eepromdirectory.c:19]: See above

File: middleware/src/eepromdirectory.c [hunk 13]
```diff
@@ -1,6 +1,33 @@
...
+
+    // calculating the address through the logic defined in shepherd
+    while (eeprom_data[i].id != NULL)
+    {
```
Comment (nwdepatie) [inline middleware/src/eepromdirectory.c:23]: nit: curly bracket formatting not following convention

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Prob won't build, check my comments

File: middleware/src/eepromdirectory.c [hunk 14]
```diff
@@ -10,21 +10,18 @@
...
-        offset += eeprom_data[i-1].size;
-        eeprom_data[i].address = offset;
-        i++;
-    }
+    for (int i = 1; i < array_size; i++) {
+        eeprom_data[i].id = partition_list[i].id;
+        eeprom_data[i].size;
+
+        // calculating the address through the logic defined in shepherd
+        eeprom_data[i].address = eeprom_data[i-1].size - eeprom-data[i-1].address;
```
Comment (nwdepatie) [inline middleware/src/eepromdirectory.c:24]: Uh I think that this should actually be + not -

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Incorrect operation (I think my bad cause I wrote it), also didn't address my comments in the header

**Review by nwdepatie — APPROVED** [review]
lgtm
